app/src/main/java/com/ichi2/apisample/AnkiDroidHelper.java

```diff
@@ -224,23 +224,36 @@ public Long addNote(long modelId, long deckId, Map<String, String> data, Set<Str
         return getApi().addNote(modelId, deckId, result, tags);
     }
 
+    @SuppressWarnings("unchecked")
     public LinkedList<Map<String, String>> findNotes(long modelId, Map<String, String> data)
             throws InvalidAnkiDatabase_fieldAndFieldNameCountMismatchException {
+        return findNotes(modelId, new Map[]{data});
+    }
+
+    public LinkedList<Map<String, String>> findNotes(long modelId, Map<String, String>[] dataSet)
+            throws InvalidAnkiDatabase_fieldAndFieldNameCountMismatchException {
         String[] fieldNames = getFieldList(modelId);
-        StringBuilder fieldsAggregated = new StringBuilder();
 
-        for (String fieldName : fieldNames) {
-            if (fieldsAggregated.length() > 0) {
-                fieldsAggregated.append(FLDS_SEPARATOR);
+        StringBuilder dataCondition = new StringBuilder();
+        for (Map<String, String> data : dataSet) {
+            if (dataCondition.length() > 0) {
+                dataCondition.append(" or ");
+            }
+            StringBuilder fieldsAggregated = new StringBuilder();
+            for (String fieldName : fieldNames) {
+                if (fieldsAggregated.length() > 0) {
+                    fieldsAggregated.append(FLDS_SEPARATOR);
+                }
+                final String value = data.containsKey(fieldName) && !data.get(fieldName).isEmpty()
+                        ? data.get(fieldName)
+                        : "%";
+                fieldsAggregated.append(value);
             }
-            final String value = data.containsKey(fieldName) && !data.get(fieldName).isEmpty()
-                    ? data.get(fieldName)
-                    : "%";
-            fieldsAggregated.append(value);
+            dataCondition.append(String.format(Locale.US, "%s like \"%s\"", FlashCardsContract.Note.FLDS, fieldsAggregated.toString()));
         }
 
-        String selection = String.format(Locale.US, "%s=%d and %s like \"%s\"",
-                FlashCardsContract.Note.MID, modelId, FlashCardsContract.Note.FLDS, fieldsAggregated.toString());
+        String selection = String.format(Locale.US, "%s=%d and (%s)",
```

Comment:
[Just as notice, needs more testing first] Not sure if the selection length is limited somehow, but it looks like selection may be long in some cases, what may affect performance (in theory).

Reply:
Nice catch. It does appear to affect performance when searching for cards with any parameters (all checkboxes are selected). Although I don't think there is a way to optimize the query. Other solutions that come to mind are throttling and maybe querying by non-permutable fields (everything except octave, note, and interval) followed by in-memory filtering by them.

Reply:
Created separate issue (#48)